src/ripple/protocol/ErrorCodes.h

```diff
@@ -27,82 +27,110 @@ namespace ripple {
 
 // VFALCO NOTE These are outside the RPC namespace
 
+// NOTE: Although the precise numeric values of these codes were never
+// intended to be stable, several API endpoints include the numeric values.
+// Some users came to rely on the values, meaning that renumbering would be
+// a breaking change for those users.
+//
+// We therefore treat the range of values as stable although they are not
+// and are subject to change.
+//
+// Please only append to this table. Do not "fill-in" gaps and do not re-use
+// or repurpose error code values.
 enum error_code_i
 {
-    rpcUNKNOWN = -1,     // Represents codes not listed in this enumeration
+    // -1 represents codes not listed in this enumeration
+    rpcUNKNOWN               = -1,
 
-    rpcSUCCESS = 0,
+    rpcSUCCESS               = 0,
 
-    rpcBAD_SYNTAX,  // Must be 1 to print usage to command line.
-    rpcJSON_RPC,
-    rpcFORBIDDEN,
-
-    // Error numbers beyond this line are not stable between versions.
-    // Programs should use error tokens.
+    rpcBAD_SYNTAX            = 1,
+    rpcJSON_RPC              = 2,
+    rpcFORBIDDEN             = 3,
 
     // Misc failure
-    rpcNO_PERMISSION,
-    rpcNO_EVENTS,
-    rpcTOO_BUSY,
-    rpcSLOW_DOWN,
-    rpcHIGH_FEE,
-    rpcNOT_ENABLED,
-    rpcNOT_READY,
-    rpcAMENDMENT_BLOCKED,
+    // unused                  4,
+    // unused                  5,
+    rpcNO_PERMISSION         = 6,
```

Comment:
So users were expecting `rpcNO_PERMISSION` to be 6, but we were returning 4? Just want to clarify that this change is to fix an issue that was caused by adding/removing error codes.

Reply:
That's basically correct.  The specific complaint was that the error code for `actNotFound` changed from 19 to 15.  But the machinery is identical.

The reason that change occurred is that In a previous commit I had removed a bunch of RPC error messages that were no longer being used.  That's what caused the issue.  To recover I pulled up an old version of rippled and inserted `// unused` comments where the old errors and their codes had been.  The enum values now match the values found in this commit: https://github.com/ripple/rippled/commit/36d6758945954fab757b024b2afc8c3a5ad76919